KAFKA-5556 : KafkaConsumer.commitSync throws IllegalStateException: A…

umesh9794 opened the pull request:
This PR makes `commitOffsetsSync` method check whether future is completed after client's poll or not. 

Tests: All existing tests especially "`testCommitOffsetSyncCallbackWithNonRetriableException`" is passed. Not sure if we need to add any dedicated tests for this minor change. 

Awaiting your review comments.

umesh9794 committed:
KAFKA-5556 : KafkaConsumer.commitSync throws IllegalStateException: Attempt to retrieve exception from future which hasn't failed

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5901/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5886/
Test FAILed (JDK 8 and Scala 2.12).

ijuma commented:
Thanks for the PR. We should add a test, yes.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -617,7 +618,7 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
             now = time.milliseconds();
             remainingMs = timeoutMs - (now - startMs);
-        } while (remainingMs > 0);
+        } while (remainingMs > 0 && !isFutureCompleted);
```
I don't think this fixes the issue. The issue is the following:

```java
if (!future.isRetriable())
    throw future.exception();
```

It should be:

```java
if (future.failed() && !future.isRetriable())
    throw future.exception();
```

It looks like this may not be the only place where we call `isRetriable` without checking if the future failed. We should consider changing this method to avoid this error altogether. One option would be to replace `succeeded` and `failed` with a status method returning an enum with 3 states: `SUCCEEDED`, `FAILED`, `RETRY`. In most cases, we have to handle these 3 states differently.

@hachikuji, what do you think?

umesh9794 replied:
Thanks for pointing the correct fix and review comments @ijuma. Sure, I will update the PR soon after adding the test.

ijuma replied:
@hachikuji thoughts on this?

hachikuji replied:
Another way to see this bug is a failure to ensure completion of the future. Had we done so, then we could have skipped the `failed` check. This is why it worked prior to the patch which added the timeout. The pattern should really be something like this:
```java
if (future.isDone()) {
  if (future.succeeded()) {
    // handle success
  } else {
    // handle failure
  }
}
```
I guess one benefit of the enum approach is that it forces you to ensure completion prior to checking any of the possible results. That said, I'm a bit more inclined to remove the `isRetriable` method and leave it to the caller to determine what is and is not retriable. Then the request future only has two completion states.

ijuma replied:
OK, I filed https://issues.apache.org/jira/browse/KAFKA-5606 to track this discussion separately from this PR.

umesh9794 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-5556

umesh9794 committed:
KAFKA-5556 : KafkaConsumer.commitSync throws IllegalStateException: Attempt to retrieve exception from future which hasn't failed

umesh9794 commented:
@ijuma I have updated the PR with review comments and added the test as well. Please review and let me know if unit test is sufficient for this change.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6001/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5986/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji reviewed (approved):
Good catch, LGTM. Looks like we missed this when we added the timeout here: https://github.com/apache/kafka/commit/f72203ee9223d3b724ee67bdad9912612dd72f63.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1287,6 +1287,16 @@ public void testCommitAsyncNegativeOffset() {
     }
 
+    @Test(expected = CommitFailedException.class)
+    public void testCommitOffsetSyncCallbackWithoutNonRetriableException() {
```
Sorry, jumped the gun on my +1. I don't think this is quite right. We need to hit the case when the request times out, right?

umesh9794 replied:
Thanks for the review @hachikuji , in the test we need to make sure that future gets failed (due to any possible error, in this case it is UNKNOWN_MEMBER_ID) and offset commit sync request throws the appropriate exception when it discovers that the future is failed and it is non-retriable also. That is why we added below to prevent the message mentioned in the JIRA : 

`            if (future.failed() && !future.isRetriable())
`

I see that currently we don't try to commit the offsets if timeout occurs. Please let me know if I am missing something. 

@ijuma please correct me if I am wrong.

ijuma replied:
@umesh9794 I think Jason is saying that the following line must return a Future that is not complete for this test to verify the fix:

```java
client.poll(future, remainingMs);
```

If that method returns a completed `Future`, then the test will pass without the fix in this PR. You can verify this by removing the fix and running the test. Does it pass, in that case?

umesh9794 replied:
Ah, I see that now after quickly running the test without fix. Thanks for the clear explanation @ijuma. However I am not sure how to create the condition which causes the future not to complete from `client.poll` method. Can you please add some pointers for this?

ijuma replied:
You may need to change `MockClient.prepareResponse` so that you can pass a requestDurationMs, which you can store in FutureResponse. In `send`, you can compute `ClientResponse.receiveTimeMs` to include `requestDurationMs`. In `poll`, you only return responses where `now + timeoutMs => receiveTimeMs`. Does this make sense?

umesh9794 replied:
@ijuma , it took me a while to absorb the reason behind the suggested changes (Please consider that I am a newbie to the related classes). Although I made those changes, still my test has some glitches as it forever remains in the loop of `ConsumerNetworkClient.poll`. Can I have your pointers to fix that, please ?

ijuma replied:
Sure, but I'll need to run it. I probably won't have time before tomorrow.

umesh9794 replied:
Sure, Thank you :)

umesh9794 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-5556

umesh9794 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-5556

umesh9794 committed:
KAFKA-5556 : KafkaConsumer.commitSync throws IllegalStateException: Attempt to retrieve exception from future which hasn't failed

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6166/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6150/
Test FAILed (JDK 8 and Scala 2.12).

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1287,6 +1287,16 @@ public void testCommitAsyncNegativeOffset() {
     }
 
+    @Test(expected = CommitFailedException.class)
+    public void testCommitOffsetSyncCallbackWithCommitFailedExceptionForTimeout() {
```
I'm not quite sure how this is supposed to work. The case we're trying to hit is when `client.poll(future, timeout)` returns without the future being completed. A simple way to check this path is something like this:
```java
       client.prepareResponse(groupCoordinatorResponse(node, Errors.NONE));
       coordinator.ensureCoordinatorReady();
       assertFalse(coordinator.commitOffsetsSync(Collections.singletonMap(t1p, new OffsetAndMetadata(100L)), 0));
 ```
You can confirm that this test fails without the fix above.

umesh9794 replied:
Ah, the best way to make sure that future doesn't complete is not to prepare `offsetCommitResponse`. Sorry I got that bit late. Thank you @hachikuji for pointing that. I tested it with and without the fix and it works as expected. Will be sending the updated PR soon.

ijuma replied:
My bad for suggesting an unnecessarily complex solution.

umesh9794 replied:
@ijuma , thanks for that because that made me to read and understand the code of `MockClient`. So at this point should I revert the changes of `MockClient` for this PR? Please  let me know your thoughts.

umesh9794 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-5556

umesh9794 committed:
KAFKA-5556 : KafkaConsumer.commitSync throws IllegalStateException: Attempt to retrieve exception from future which hasn't failed

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6192/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6176/
Test FAILed (JDK 8 and Scala 2.12).

umesh9794 committed:
Revert: MockClient Changes

umesh9794 commented:
Please review the updated PR.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6180/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6196/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji reviewed (approved):
LGTM. I will merge to trunk and 0.11.0.
